src/printer.js

```diff
@@ -1994,6 +1955,71 @@ function printMemberChain(node, options, print) {
   }
 }
 
+function printJSXElement(path, options, print) {
+  const n = path.getValue();
+  var openingLines = path.call(print, "openingElement");
+
+  let elem;
+  if (n.openingElement.selfClosing) {
+    assert.ok(!n.closingElement);
+
+    elem = openingLines;
+  } else {
+    var children = [];
+
+    path.map(
+      function(childPath) {
+        var child = childPath.getValue();
+
+        if (
+          namedTypes.Literal.check(child) && typeof child.value === "string"
+        ) {
+          if (/\S/.test(child.value)) {
+            const beginBreak = child.value.match(/^\s*\n/);
+            const endBreak = child.value.match(/\n\s*$/);
+
+            children.push(
+              beginBreak ? hardline : "",
+              child.value.replace(/^\s+|\s+$/g, endBreak ? "" : " "),
+              endBreak ? hardline : ""
+            );
+          } else if (/\n/.test(child.value)) {
+            children.push(hardline);
+          }
+        } else {
+          children.push(print(childPath));
+        }
+      },
+      "children"
+    );
+
+    var mostChildren = children.slice(0, -1);
+    var closingLines = path.call(print, "closingElement");
+    elem = concat([
+      openingLines,
+      indent(options.tabWidth, concat(mostChildren)),
+      util.getLast(children) || "",
+      closingLines
+    ]);
+  }
+
+  const parent = path.getParentNode();
+  if (!parent || parent.type === "JSXElement" || parent.type === "ExpressionStatement") {
+    return elem;
+  }
+
+  const groups = [];
+  if (!hasHardLine(elem)) groups.push(elem);
+  groups.push(concat([
+    "(",
+    indent(options.tabWidth, concat([ hardline, elem ])),
+    hardline,
+    ")",
+  ]));
+
+  return conditionalGroup(groups);
```

Comment:
I don't think you need `conditionalGroup` here. Use `conditionalGroup` only as a last resort :) (I plan on writing more documentation about all this)

`conditionalGroup` is only needed when you really need manual control over how things are displayed when they don't fit. The majority of use cases fits into these 2 categories though:

1. You want to break something across lines and indent if it doesn't fit (this assumes nothing inside of it has "hard" lines, meaning all contents try to fit on the same line)
2. You want to break something across lines if any of its children are broken across lines (they have "hard" lines)

`group` solves #1 easily: use `indent` and `line` and the printer, when the group doesn't fit, will output newlines where `line` was used and it will be indented an extra level.

I made `multilineGroup` for #2 because it's a very common use case. It will see if any of its contents have hard lines, and if they do, the group will break. Hard lines are lines that emit a newline no matter what. The reason we don't have to worry about normal breaking is because this group is always larger than it's contents -- if anything inside of it naturally breaks, this group will already have broken as well.

There are some rare cases where you want to let something inside of you break, but not break the group, and control indentation rules in each specific case. That's what `conditionalGroup` is for, but we should use it rarely as it has a performance cost.

There's one last character in this cast: `ifBreak`. This is a very simple construct that does what it says: if the group break, include this content. You can say `ifBreak(",")` and it will only include the comma if the group has broken. The solves some very simple use cases.

Will all that said, I think this should work:

```js
multilineGroup(concat([
  ifBreak("("),
  indent(options.tabWidth, concat([ softline, elem ])),
  softline,
  ifBreak(")"),
]))
```